gcc/rust/hir/rust-ast-lower-stmt.cc

```diff
@@ -68,21 +69,24 @@ ASTLoweringStmt::visit (AST::LetStmt &stmt)
 {
   HIR::Pattern *variables
     = ASTLoweringPattern::translate (stmt.get_pattern (), true);
-  HIR::Type *type = stmt.has_type ()
-		      ? ASTLoweringType::translate (stmt.get_type ())
-		      : nullptr;
-  HIR::Expr *init_expression
-    = stmt.has_init_expr () ? ASTLoweringExpr::translate (stmt.get_init_expr ())
-			    : nullptr;
+
+  auto type
+    = stmt.has_type () ? tl::optional<std::unique_ptr<Type>> (
+	std::unique_ptr<Type> (ASTLoweringType::translate (stmt.get_type ())))
+		       : tl::nullopt;
+  auto init_expression
+    = stmt.has_init_expr ()
+	? tl::optional<std::unique_ptr<Expr>> (std::unique_ptr<HIR::Expr> (
+	  ASTLoweringExpr::translate (stmt.get_init_expr ())))
+	: tl::nullopt;
```

Comment:
those calls are really rough. If it works I'd switch it to this, or at least remove the `auto` and put the `tl::optional` on the left side.
```suggestion
  tl::optional<std::unique_ptr<Type>> type
    = stmt.has_type () ? std::unique_ptr<Type> (ASTLoweringType::translate (stmt.get_type ()))
		       : tl::nullopt;
  tl::optional<std::unique_ptr<HIR::Expr>> init_expression
    = stmt.has_init_expr ()
	? std::unique_ptr<HIR::Expr> (
	  ASTLoweringExpr::translate (stmt.get_init_expr ()))
	: tl::nullopt;
```

Reply:
I tried but GCC 4.8 and clang on MacOS are complaining.